ospackage/recipes/artifacts.go

```diff
@@ -0,0 +1,186 @@
+	}
+
+	return localNames, nil
+}
+
+func fetchArtifact(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, directory string) (string, error) {
+	localPath := path.Join(directory, artifact.Id)
+	uri, err := url.Parse(artifact.Uri)
+	if err != nil {
+		return "", fmt.Errorf("Could not parse url %q for artifact %q", artifact.Uri, artifact.Id)
+	}
+
+	var reader io.ReadCloser
+	switch strings.ToLower(uri.Scheme) {
+	case "gcs":
+		reader, err = fetchWithGCS(ctx, uri.Host, uri.Path, uri.Fragment)
+		defer reader.Close()
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q from GCS: %v", artifact.Id, err)
+		}
+	case "https", "http":
+		reader, err = fetchWithHTTP(ctx, artifact.Uri)
+		defer reader.Close()
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q with http or https: %v", artifact.Id, err)
+		}
+	default:
+		return "", fmt.Errorf("artifact %q has unsupported protocol %s", artifact.Id, uri.Scheme)
+	}
+
+	err = downloadStream(reader, artifact.Checksum, localPath)
+	if err != nil {
+		return "", err
+	}
+	return localPath, nil
+}
+
+func fetchWithGCS(ctx context.Context, bucket, path, generation string) (io.ReadCloser, error) {
+	client, err := newStorageClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create storage client: %v", err)
+	}
+	defer client.Close()
+
+	oh := client.Bucket(bucket).Object(path)
+	if generation != "" {
+		generationNumber, err := strconv.ParseInt(generation, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't parse gcs generation number %q", generation)
+		}
+		oh = oh.Generation(generationNumber)
+	}
+
+	r, err := oh.NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func fetchWithHTTP(ctx context.Context, uri string) (io.ReadCloser, error) {
+	resp, err := newHTTPClient().Get(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("got http status %d when attempting to download artifact", resp.StatusCode)
+	}
+
+	return resp.Body, nil
+}
+
+func downloadStream(r io.Reader, checksum string, localPath string) error {
+	file, err := createFile(localPath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	hasher := sha256.New()
+	_, err = io.Copy(io.MultiWriter(file, hasher), r)
+	if err != nil {
+		return err
+	}
+	computed := fmt.Sprintf("%64x", hasher.Sum(nil))
+	if checksum != "" && !strings.EqualFold(checksum, computed) {
+		return fmt.Errorf("got %q for checksum, expected %q", computed, checksum)
+	}
+	return nil
+}
+
+// normPath transforms a windows path into an extended-length path as described in
+// https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx#maxpath
+// when not running on windows it will just return the input path. Copied from
+// https://github.com/google/googet/blob/master/oswrap/oswrap_windows.go
+func normPath(path string) (string, error) {
+	if runtime.GOOS != "windows" {
```

Comment:
can just do the runtime check in caller to determine whether normPath is needed

Reply:
The reason I chose the current option is to make is so we don't need that same check before every call to normpath. Many of the steps will need to use the os library so this function will need to be called a decent amount in the steps.